### PR TITLE
- Function API: Add JS binding for OCSetDeviceId

### PR DESCRIPTION
**Pull request by Touyote**
Signed-off-by: GASQ Romain <r.gasq@samsung.com>

### CODE REVIEW

File: src/functions/simple.cc [hunk 1]
```diff
@@ -102,6 +102,32 @@ napi_value bind_OCGetServerInstanceIDString(napi_env env,
...
 }
 
+napi_value bind_OCSetDeviceId(napi_env env, napi_callback_info info)
+{
+    OCUUIdentity    *ocid = NULL;
+    OCStackResult   res = OC_STACK_INVALID_PARAM;
+    char            byte[2] = {0};
+
+    J2C_DECLARE_ARGUMENTS(env, info, 1);
+    J2C_GET_STRING_TRACKED_JS_THROW(uuid, env, arguments[0], true, "uuid");
```
Comment (gabrielschulhof) [inline src/functions/simple.cc]: Why not simply use `false` and then the function will return if `uuid` is `NULL`, and the if-statement below is unnecessary?

File: src/functions/simple.cc [hunk 2]
```diff
@@ -102,6 +102,32 @@ napi_value bind_OCGetServerInstanceIDString(napi_env env,
...
+    if (!uuid)
+        C2J_SET_RETURN_VALUE(env, info, double, ((double)res));
+    ocid = (OCUUIdentity*)malloc(sizeof(*ocid));
+    if (!ocid)
```
Comment (gabrielschulhof) [inline src/functions/simple.cc]: This if-statement needs to be replaced with
```C++
JS_ASSERT(ocid != 0, "Failed to allocate identity", THROW_BODY(env, 0));
```

File: src/functions/simple.cc [hunk 3]
```diff
@@ -102,6 +102,32 @@ napi_value bind_OCGetServerInstanceIDString(napi_env env,
...
+        C2J_SET_RETURN_VALUE(env, info, double, ((double)OC_STACK_NO_MEMORY));
+    memset(ocid, 0, UUID_IDENTITY_SIZE);
+    for (unsigned int i = 0, k = 0, len = strlen(uuid); i < len; i += 2, ++k) {
+        if (uuid[i] == '-')
+            ++i;
+        strncpy(&byte[0], &uuid[i], 2);
+        ocid->id[k] = strtol(byte, NULL, 16);
+    }
```
Comment (gabrielschulhof) [inline src/functions/simple.cc]: Please use [`OCConvertStringToUuid`](https://api-docs.iotivity.org/latest-c/ocrandom_8h.html#a8a6aa50346f1f027b0af574f8dd3e860) instead.

File: src/functions/simple.cc [hunk 4]
```diff
@@ -102,6 +102,32 @@ napi_value bind_OCGetServerInstanceIDString(napi_env env,
...
+    res = OCSetDeviceId(ocid);
+    if (ocid)
```
Comment (gabrielschulhof) [inline src/functions/simple.cc]: If you convert the null-check above into an assertion, you can unconditionally free `ocid` here.

**Comment by gabrielschulhof** [comment]
@Touyote there is [a test](https://github.com/intel/iotivity-node/blob/master/tests/tests/Interactive%20Server%20Startup.js#L30) that gives you an example of how to launch a child process which runs a test.

**Comment by gabrielschulhof** [comment]
Oh, and if you do check the behaviour in persistent storage, in your test, you'll also want to precede the call to `OCInit()` with
```JS
result = iotivity.OCRegisterPersistentStorageHandler( require( "../../../lib/StorageHandler" )() );
```
where the path to `StorageHandler.js` is adjusted to match the `lib` directory off the repo root.

**Comment by Touyote** [comment]
Hello @gabrielschulhof ,
Thanks for having taken your time to check my code, I will change the necessary and adding the test for sure.
Good things to check the persistent storage, actually in my test case I develop a simulator of an OCF Device which is based on the [multi-server-server2.js](https://github.com/intel/iotivity-node/blob/master/js/multi-server-server2.js) example.
I don't use a persistent storage at all, it is actually managed by an external component.

**Comment by gabrielschulhof** [comment]
@Touyote you need not close the PR. Just push the changes to this branch and I will look again 🙂

**Comment by gabrielschulhof** [comment]
@Touyote it looks like the semantics of `OCSetDeviceId()` are such that it is intended to set the device ID in the persistent storage.

Why not
```C++
memset(&ocid, 0, sizeof(ocid));
```
I'd make the message "Unable to convert UUID string to OCUUIDIdentity".

Please don't forget to re-open the PR!

**Comment by gabrielschulhof** [comment]
Device discovery always works because, in the absence of a device UUID, iotivity generates one at runtime. If there is no persistent storage, the UUID will be different every time the device starts.

**Comment by Touyote** [comment]
Hello @gabrielschulhof,

First of all I'm unable to re-push that P.R. I will send the new one once I finish to test certain things & prepare your simple unit test. I was switching between two fork of your project and unfortunately I 
 had to delete the repository which I used for generated the P.R....

Secondly about the UUID generation, yes is true, it's always generated when it's not define.
However, I have tested on different platform, I succeed even without a persistent storage to set the device UUID on all boards with a simple test which launch directly the OCF server. But when I used mocha to launch the server I'm not able to set it on my board.

Finally, another thing appeared, I'm not able to see the call of my callback for each OCF endpoint registered with OCCreateResource on a platform x86 & arm 64. It's weird because I'm able to do it with my arm32.

**Comment by gabrielschulhof** [comment]
@Touyote NP! Let's get the binding working and at least a basic test for it, and we can worry about testing various use cases later.

That *is* weird about the entity handler.

**Pull request closed without merging**